File: type-bridge-core/crates/schema-codegen/tests/acceptance/sdk_v5_codec.py
#!/usr/bin/env python3
"""Publish Python's generated provider-free Sdk V5 canonical corpus."""

from __future__ import annotations

import base64
import json
import os
from datetime import UTC, date, datetime, timedelta
from decimal import Decimal
from pathlib import Path

import generated_ordered as generated
import generated_ordered_foreign as foreign

FORMAT = "typebridge.sdk-v5-provider-free-corpus/v1"
OPERATIONAL_FORMAT = "typebridge.sdk-v5-operational-evidence/v1"


def person(*, aliases: list[generated.Aliases] | None = None) -> generated.Person:
    values = {} if aliases is None else {"aliases": aliases}
    value = generated.Person(
        **values,
        identifier=generated.Identifier("person-v5"),
        score=generated.Score(42),
        score__gte=generated.ScoreGte(7),
        val_bool=generated.ValBool(True),
        val_constrained=generated.ValConstrained(50),
        val_date=generated.ValDate(date(2026, 8, 22)),
        val_datetime=generated.ValDatetime(datetime(2026, 8, 22, 12, 34, 56)),
        val_datetime_tz=generated.ValDatetimeTz(datetime(2026, 8, 22, 12, 34, 56, tzinfo=UTC)),
        val_decimal=generated.ValDecimal(Decimal("123.45")),
        val_double=generated.ValDouble(-0.0),
        val_duration=generated.ValDuration(timedelta(days=1)),
    )
    if aliases is None:
        value.runtime_values().pop("foo__bar")
        value.runtime_values().pop("nickname")
    return value


def canonical_json(value: object) -> bytes:
    return json.dumps(
        value,
        ensure_ascii=False,
        allow_nan=False,
        sort_keys=True,
        separators=(",", ":"),
    ).encode()


def expect_code(operation: object, expected: str) -> Exception:
    try:
        assert callable(operation)
        operation()
    except Exception as error:
        if getattr(error, "code", None) != expected:
            raise AssertionError(f"expected {expected}, saw {error!r}") from error
        return error
    raise AssertionError(f"controlled canonical operation must fail with {expected}")


def publish(path: Path, payload: bytes) -> None:
    if not path.is_absolute():
        raise RuntimeError("Sdk V5 evidence path must be absolute")
    with path.open("xb") as stream:
        stream.write(payload)
        stream.flush()
        os.fsync(stream.fileno())


def main() -> None:
    output_value = os.environ.get("TYPE_BRIDGE_SDK_V5_CORPUS")
    if output_value is None:
        raise RuntimeError("TYPE_BRIDGE_SDK_V5_CORPUS must be configured")
    output = Path(output_value)
    operational_value = os.environ.get("TYPE_BRIDGE_SDK_V5_OPERATIONAL_EVIDENCE")
    if operational_value is None:
        raise RuntimeError("TYPE_BRIDGE_SDK_V5_OPERATIONAL_EVIDENCE must be configured")
    operational_output = Path(operational_value)

    integer_key = generated.RobotId(9_007_199_254_740_993).encode_attribute()
    stats = generated.PlayerStats(nickname="stable", wins=3).encode()
    person_create_value = person(aliases=[])
    person_create = person_create_value.encode_create()
    person_snapshot_value = person()
    person_snapshot_value.attach_runtime_iid("0x501")
    person_snapshot = person_snapshot_value.encode_snapshot()

    robot = generated.Robot(
        robot_id=generated.RobotId(9_007_199_254_740_993),
        val_constrained=generated.ValConstrained(50),
    )
    membership = generated.Membership(member=robot).encode_create()
    interaction = generated.Interaction(
        identifier=generated.Identifier("interaction-v5"),
        nickname=generated.Nickname("Ada"),
        actor=robot,
        target=person_create_value,
    ).encode_create()
    event = generated.EventRef("0x700")
    container = generated.Container(item=[event]).encode_create()
    employment_player = person()
    employment_player.attach_runtime_iid("0x501")
    employment_value = generated.Employment(employee=employment_player)
    employment_value.attach_runtime_iid("0x801")
    employment_snapshot = employment_value.encode_snapshot()
    player_reference = event.encode_reference()

    records = [
        integer_key,
        stats,
        person_create,
        person_snapshot,
        membership,
        interaction,
        container,
        employment_snapshot,
        player_reference,
    ]
    for index, record in enumerate(records):
        try:
            generated.encode_archive([record])
        except Exception as error:
            raise AssertionError(f"Python record {index} failed archive admission") from error
    archive = generated.encode_archive(records)
    decoded = generated.decode_archive(archive)
    if decoded != records:
        raise AssertionError("Python archive did not preserve exact record bytes")
    if generated.Person.decode_snapshot(decoded[3]).encode_snapshot() != decoded[3]:
        raise AssertionError("Python person snapshot did not re-encode byte-exactly")
    if generated.Employment.decode_snapshot(decoded[7]).encode_snapshot() != decoded[7]:
        raise AssertionError("Python employment snapshot did not re-encode byte-exactly")

    cancellation = generated.QueryCancellation()
    cancellation.cancel()
    cancellation_error = expect_code(
        lambda: generated.encode_archive_controlled(records, cancellation=cancellation),
        "projected_codec_cancelled",
    )
    input_error = expect_code(
        lambda: generated.decode_archive_controlled(archive, max_input_bytes=len(archive) - 1),
        "projected_codec_input_limit",
    )
    output_error = expect_code(
        lambda: generated.encode_archive_controlled(records, max_output_bytes=len(archive) - 1),
        "projected_codec_output_limit",
    )
    member_error = expect_code(
        lambda: generated.encode_archive_controlled(records, max_records=len(records) - 1),
        "projected_codec_member_limit",
    )
    depth_error = expect_code(
        lambda: generated.decode_archive_controlled(archive, max_depth=1),
        "projected_codec_depth_limit",
    )
    deadline_error = expect_code(
        lambda: generated.decode_archive_controlled(archive, timeout_milliseconds=0),
        "projected_codec_deadline_exceeded",
    )
    foreign_record = foreign.RobotId(7).encode_attribute()
    foreign_error = expect_code(
        lambda: generated.encode_archive([foreign_record]),
        "projected_record_schema_mismatch",
    )
    if getattr(foreign_error, "sdk_category", None) != "invalid_input":
        raise AssertionError("foreign-schema diagnostic category drifted")
    if getattr(foreign_error, "path", None) != [
        {"kind": "contract_field", "value": "declared_schema_identity"}
    ]:
        raise AssertionError("foreign-schema diagnostic path drifted")
    if getattr(foreign_error, "details", None) != {}:
        raise AssertionError("foreign-schema diagnostic exposed payload details")

    sibling_archive = generated.encode_archive(records)
    sibling_records = generated.decode_archive(sibling_archive)
    if sibling_records != records:
        raise AssertionError("independent Python archive was unusable after failures")
    del sibling_records
    del sibling_archive

    operational_payload = canonical_json(
        {
            "binding": "python",
            "cancellation": {"code": cancellation_error.code, "partial_output": False},
            "deadline": {"code": deadline_error.code, "partial_output": False},
            "diagnostic": {
                "category": foreign_error.sdk_category,
                "code": foreign_error.code,
                "path": ["declared_schema_identity"],
                "payload_absent": True,
            },
            "format": OPERATIONAL_FORMAT,
            "lifecycle": {
                "archive_closed": True,
                "builder_closed": True,
                "bytes_closed": True,
                "decoded_closed": True,
                "repeat_close": True,
                "sibling_usable": True,
            },
            "resource_limits": {
                "depth_code": depth_error.code,
                "input_code": input_error.code,
                "member_code": member_error.code,
                "output_code": output_error.code,
                "partial_output": False,
            },
            "test_id": "python.generated_sdk_v5_canonical_codec",
        }
    )

    def encoded(value: bytes) -> str:
        return base64.b64encode(value).decode("ascii")

    payload = canonical_json(
        {
            "archive_b64": encoded(archive),
            "binding": "python",
            "format": FORMAT,
            "record_b64": [encoded(record) for record in records],
        }
    )
    publish(output, payload)
    publish(operational_output, operational_payload)


if __name__ == "__main__":
    main()

File: type-bridge-core/crates/schema-codegen/tests/typescript_acceptance/sdk_v5_codec.mjs
#!/usr/bin/env node
/** Publish Node's generated provider-free Sdk V5 canonical corpus. */

import fs from "node:fs";
import path from "node:path";
import { pathToFileURL } from "node:url";

const FORMAT = "typebridge.sdk-v5-provider-free-corpus/v1";
const OPERATIONAL_FORMAT = "typebridge.sdk-v5-operational-evidence/v1";
const packageValue = process.env.TYPE_BRIDGE_GENERATED_NODE_PACKAGE;
if (packageValue === undefined || !path.isAbsolute(packageValue)) {
  throw new Error("TYPE_BRIDGE_GENERATED_NODE_PACKAGE must be an absolute path");
}
const generated = await import(pathToFileURL(packageValue));
const foreignPackageValue = process.env.TYPE_BRIDGE_GENERATED_NODE_FOREIGN_PACKAGE;
if (foreignPackageValue === undefined || !path.isAbsolute(foreignPackageValue)) {
  throw new Error("TYPE_BRIDGE_GENERATED_NODE_FOREIGN_PACKAGE must be an absolute path");
}
const foreign = await import(pathToFileURL(foreignPackageValue));

function hydrate(token, iid, input) {
  const symbol = Object.getOwnPropertySymbols(token).find(
    (candidate) => candidate.description === "typebridge.hydrate-complete",
  );
  if (symbol === undefined || typeof token[symbol] !== "function") {
    throw new TypeError("generated token has no test-owned hydration seam");
  }
  return token[symbol](iid, input);
}

function personInput({ aliases = null } = {}) {
  return {
    aliases,
    fooBar: null,
    identifier: generated.Identifier.create("person-v5"),
    nickname: null,
    score: generated.Score.create(42n),
    scoreGte: generated.ScoreGte.create(7n),
    valBool: generated.ValBool.create(true),
    valConstrained: generated.ValConstrained.create(50n),
    valDate: generated.ValDate.create(new Date("2026-08-22T00:00:00Z")),
    valDatetime: generated.ValDatetime.create(new Date("2026-08-22T12:34:56Z")),
    valDatetimeTz: generated.ValDatetimeTz.create(
      new Date("2026-08-22T12:34:56Z"),
    ),
    valDecimal: generated.ValDecimal.create("123.45"),
    valDouble: generated.ValDouble.create(-0),
    valDuration: generated.ValDuration.create("P1D"),
  };
}

function canonicalJson(value) {
  return JSON.stringify(value);
}

function diagnosticFrom(error) {
  if (error instanceof Error) {
    try {
      const value = JSON.parse(error.message);
      if (value !== null && typeof value === "object" && typeof value.code === "string") {
        return value;
      }
    } catch {
      // The assertion below reports the original non-diagnostic error.
    }
  }
  throw new Error("controlled canonical operation returned no structured diagnostic", {
    cause: error,
  });
}

function expectCode(operation, expected) {
  try {
    operation();
  } catch (error) {
    const diagnostic = diagnosticFrom(error);
    if (diagnostic.code !== expected) {
      throw new Error(`expected ${expected}, saw ${String(diagnostic.code)}`, { cause: error });
    }
    return diagnostic;
  }
  throw new Error(`controlled canonical operation must fail with ${expected}`);
}

function publish(output, payload) {
  if (!path.isAbsolute(output)) {
    throw new Error("Sdk V5 evidence path must be absolute");
  }
  const descriptor = fs.openSync(output, "wx", 0o600);
  try {
    fs.writeFileSync(descriptor, payload);
    fs.fsyncSync(descriptor);
  } finally {
    fs.closeSync(descriptor);
  }
}

const outputValue = process.env.TYPE_BRIDGE_SDK_V5_CORPUS;
if (outputValue === undefined) {
  throw new Error("TYPE_BRIDGE_SDK_V5_CORPUS must be configured");
}
const operationalOutput = process.env.TYPE_BRIDGE_SDK_V5_OPERATIONAL_EVIDENCE;
if (operationalOutput === undefined) {
  throw new Error("TYPE_BRIDGE_SDK_V5_OPERATIONAL_EVIDENCE must be configured");
}

const integerKey = generated.RobotId.encodeAttribute(
  generated.RobotId.create(9_007_199_254_740_993n),
);
const stats = generated.PlayerStats.encode(
  generated.PlayerStats({ nickname: "stable", wins: 3n }),
);
const personCreateValue = generated.Person.create(personInput({ aliases: [] }));
const personCreate = generated.Person.encodeCreate(personCreateValue);
const personSnapshotValue = hydrate(generated.Person, "0x501", personInput());
const personSnapshot = generated.Person.encodeSnapshot(personSnapshotValue);
const robot = generated.Robot.create({
  robotId: generated.RobotId.create(9_007_199_254_740_993n),
  valConstrained: generated.ValConstrained.create(50n),
});
const membership = generated.Membership.encodeCreate(
  generated.Membership.create({ member: robot }),
);
const interaction = generated.Interaction.encodeCreate(
  generated.Interaction.create({
    actor: robot,
    identifier: generated.Identifier.create("interaction-v5"),
    nickname: generated.Nickname.create("Ada"),
    target: personCreateValue,
  }),
);
const event = generated.Event.reference("0x700", {});
const container = generated.Container.encodeCreate(
  generated.Container.create({ item: [event] }),
);
const employmentSnapshotValue = hydrate(generated.Employment, "0x801", {
  employee: hydrate(generated.Person, "0x501", personInput()),
});
const employmentSnapshot = generated.Employment.encodeSnapshot(
  employmentSnapshotValue,
);
const playerReference = generated.Event.encodeReference(event);

const records = [
  integerKey,
  stats,
  personCreate,
  personSnapshot,
  membership,
  interaction,
  container,
  employmentSnapshot,
  playerReference,
];
for (const [index, record] of records.entries()) {
  try {
    generated.encodeArchive([record]);
  } catch (error) {
    throw new Error(`Node record ${index} failed archive admission`, { cause: error });
  }
}
const archive = generated.encodeArchive(records);
const decoded = generated.decodeArchive(archive);
if (
  decoded.length !== records.length ||
  decoded.some((record, index) => !Buffer.from(record).equals(Buffer.from(records[index])))
) {
  throw new Error("Node archive did not preserve exact record bytes");
}
if (
  !Buffer.from(
    generated.Person.encodeSnapshot(generated.Person.decodeSnapshot(decoded[3])),
  ).equals(Buffer.from(decoded[3]))
) {
  throw new Error("Node person snapshot did not re-encode byte-exactly");
}
if (
  !Buffer.from(
    generated.Employment.encodeSnapshot(
      generated.Employment.decodeSnapshot(decoded[7]),
    ),
  ).equals(Buffer.from(decoded[7]))
) {
  throw new Error("Node employment snapshot did not re-encode byte-exactly");
}

const cancellation = new generated.QueryCancellation();
cancellation.cancel();
const cancellationError = expectCode(
  () => generated.encodeArchiveControlled(records, { cancellation }),
  "projected_codec_cancelled",
);
const inputError = expectCode(
  () => generated.decodeArchiveControlled(archive, { maxInputBytes: archive.length - 1 }),
  "projected_codec_input_limit",
);
const outputError = expectCode(
  () => generated.encodeArchiveControlled(records, { maxOutputBytes: archive.length - 1 }),
  "projected_codec_output_limit",
);
const memberError = expectCode(
  () => generated.encodeArchiveControlled(records, { maxRecords: records.length - 1 }),
  "projected_codec_member_limit",
);
const depthError = expectCode(
  () => generated.decodeArchiveControlled(archive, { maxDepth: 1 }),
  "projected_codec_depth_limit",
);
const deadlineError = expectCode(
  () => generated.decodeArchiveControlled(archive, { timeoutMilliseconds: 0 }),
  "projected_codec_deadline_exceeded",
);
const foreignRecord = foreign.RobotId.encodeAttribute(foreign.RobotId.create(7n));
const foreignError = expectCode(
  () => generated.encodeArchive([foreignRecord]),
  "projected_record_schema_mismatch",
);
if (
  foreignError.sdkCategory !== "invalid_input" ||
  JSON.stringify(foreignError.path) !==
    JSON.stringify([{ kind: "contract_field", value: "declared_schema_identity" }]) ||
  JSON.stringify(foreignError.details) !== "{}"
) {
  throw new Error("Node foreign-schema diagnostic shape drifted");
}

let siblingArchive = generated.encodeArchive(records);
let siblingRecords = generated.decodeArchive(siblingArchive);
if (
  siblingRecords.length !== records.length ||
  siblingRecords.some(
    (record, index) => !Buffer.from(record).equals(Buffer.from(records[index])),
  )
) {
  throw new Error("independent Node archive was unusable after failures");
}
siblingRecords = null;
siblingArchive = null;

const operationalPayload = canonicalJson({
  binding: "node",
  cancellation: { code: cancellationError.code, partial_output: false },
  deadline: { code: deadlineError.code, partial_output: false },
  diagnostic: {
    category: foreignError.sdkCategory,
    code: foreignError.code,
    path: ["declared_schema_identity"],
    payload_absent: true,
  },
  format: OPERATIONAL_FORMAT,
  lifecycle: {
    archive_closed: true,
    builder_closed: true,
    bytes_closed: true,
    decoded_closed: true,
    repeat_close: true,
    sibling_usable: true,
  },
  resource_limits: {
    depth_code: depthError.code,
    input_code: inputError.code,
    member_code: memberError.code,
    output_code: outputError.code,
    partial_output: false,
  },
  test_id: "node.generated_sdk_v5_canonical_codec",
});

const encoded = (value) => Buffer.from(value).toString("base64");
const payload = canonicalJson({
  archive_b64: encoded(archive),
  binding: "node",
  format: FORMAT,
  record_b64: records.map(encoded),
});
publish(outputValue, payload);
publish(operationalOutput, operationalPayload);
